### ISSUE TITLE
Tensor Indexing Loses Shape Information

### ISSUE THREAD

**Issue by idofr** [tensorflow/tensorflow#28083]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below):  v1.12.0-9492-g2c319fb415 2.0.0-alpha0
- Python version: 3.7.3

**Describe the current behavior**
Sliced tensors (e.g. `my_tensor[:, :, :]`) lose their shape information. Using `tf.slice` does seem to work but for that I need to have all dimensions defined which is more often than not simply not the case.

**Describe the expected behavior**
Shape information is kept and adjusted to match the sliced data

**Comment by achandraa** [comment]
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by idofr** [comment]
### Code
`
import tensorflow as tf

print(tf.__version__)
inp = tf.keras.layers.Input(
    shape=[32, 32, 3],
    dtype=tf.dtypes.float32
)
print(inp)

t = inp[:, :2, :, :]
print(t)
`

### tf.1.13 output

> 1.13.1
Tensor("input_1:0", shape=(?, 32, 32, 3), dtype=float32)
Tensor("strided_slice:0", shape=(?, 2, 32, 3), dtype=float32)

### tf.2.0 output

> 2.0.0-alpha0
Tensor("input_1:0", shape=(None, 32, 32, 3), dtype=float32)
Tensor("strided_slice:0", dtype=float32)

You can see the missing shape information which makes further computations significantly more complicated

**Comment by achandraa** [comment]
Thank you for providing the code snippet. I was able to reproduce.

**Comment by shashvatshahi1998** [comment]
@achandraa  @jvishnuvardhan  can you please provide me with the file where I can contribute

**Comment by alextp** [comment]
@robieta can you take a look? I think this is due to behavior of the under-the-hood keras graph losing some shape inference data from placeholders.

**Comment by ymodak** [comment]
This is fixed latest tf 2.0 nightly version '2.0.0-dev20190906'. Thanks!
Output:
```python
Tensor("input_1:0", shape=(None, 32, 32, 3), dtype=float32)
Tensor("strided_slice:0", shape=(None, 2, 32, 3), dtype=float32)
```
